### ISSUE TITLE
Opening and closing animations of cutscene bars go by way too quickly

### ISSUE THREAD

**Issue by InfoTeddy** [Colon-D/VVVVVV-over-30-FPS#3]
The opening and closing animations of cutscene bars (i.e. when the bars come in from and go out to the left and right sides of the screen) go by way too quickly.

When I try running the game under libTAS (which I cannot do at proper 60 FPS, i.e. 1000/17 FPS, so I'm running it at 30 FPS aka 1000/34 FPS), this effect seems to disappear. Are the cutscene bar animations based on realtime or something?

This is important because most scripts directly depend on the actual _animation_ of the cutscene bars (exceptions include picking up a trinket and rescuing a crewmate in a custom level). If the animation goes by too quick, the timings of scripts are directly changed as a result.

**Comment by Colon-D** [comment]
This is hopefully fixed, the original values are now untouched.

In the original game, fadeamount was an integer, which moved at 24 pixels per frame.

I had changed this to a floating point value, which moved at 24 pixels per 1/30th of a second using deltatime.

This could make timings of fading inconsistant by 1 frame, which I did not realise would cause problems.

An even bigger problem was that I was using the wrong equation for both the new speed and for deltatime, which made them run at 110% speed. Whoops. More information is in the "Warp backgrounds of rooms do not animate while in the editor" issue, as I was fixing that when I found this out.

Fixing this equation might have been enough, but since the fading actually determines when the game transitions and is not purely visual, I have now changed this back to it's original updating code, 24 pixels per frame in the fixed render function, to make it look smooth I have added an oldfadeamount value, which is updated to fadeamount before fadeamount is changed. I now do a linear interpolation between the old and current fadeamount when rendering.

**Comment by InfoTeddy** [comment]
The cutscene bars still seem like they're going too fast. They still seem to be based on deltatime instead of the actual 30-FPS game loop.

**Comment by Colon-D** [comment]
As far I can tell it's not using deltatime at all, it was added in this commit: https://github.com/Colon-D/VVVVVV-over-30-FPS/commit/d53a5e12630e9f40b504684bcf0f9c3b1d01e7d6
reverted and changed to render via a linear interpolation in this commit: https://github.com/Colon-D/VVVVVV-over-30-FPS/commit/89ec3a368ab8e526ac1576ecea22ac24653943a4

It might be one frame too fast if it's getting updated in the post fixed render function instead of pre fixed render function.

When you say "most scripts directly depend on the actual animation", do you mean they check an in game value or are they checking a pixel on the screen? Because the only thing that should be affected is the pixel on the screen, not the in game value.

**Comment by InfoTeddy** [comment]
They directly depend on `dwgfx.cutscenebarspos`, I'm pretty sure. Take a look at the `untilbars()` command, which is the one that directly depends on it, and the command that the scripts use:
```c++
			if (words[0] == "untilbars")
			{
				if (dwgfx.showcutscenebars)
				{
					if (dwgfx.cutscenebarspos < 360)
					{
						scriptdelay = 1;
						position--;
					}
				}
				else
				{
					if (dwgfx.cutscenebarspos > 0)
					{
						scriptdelay = 1;
						position--;
					}
				}
			}
```

**Comment by Colon-D** [comment]
As far as I am aware I have never touched that value, but I'll look into where it is updated

**Comment by Colon-D** [comment]
I was confusing fade during transitions with cutscene bars, so that solves my initial confusion, I'll look directly at the cutscene bars now

**Comment by InfoTeddy** [comment]
Yes, I did notice that `fadeout()`s and `fadein()`s seemed to have gotten fixed, even though I never brought them up. Before, they were also going too fast.

**Comment by Colon-D** [comment]
So, the cutscene bars are updated in the render function, which is fine at 30fps, not so fine at 1000fps or whatever the game is running at. I'll move the cutscene bar updates into the fixed render function which is called at 30fps, and I'll add a linear interpolation during variable render function to make it look smooth.

I didn't know the game had an animation for cutscene bars, whoops

**Comment by Colon-D** [comment]
I have pushed the latest commit and released the v1.1.1 zip, which should fix the issue, they are now updated in the post fixed render function. If there are any more issues related to this (ie, off by 1 frame errors) then moving it to the pre fixed render function should fix that.

**Comment by InfoTeddy** [comment]
Yep, the cutscene bars seem to be fixed now. Thank you!
